### ISSUE TITLE
Keras model evaluate() progress bar randomly stops before 100%

### ISSUE THREAD

**Issue by ageron** [tensorflow/tensorflow#24593]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Mac OS X 10.13.6
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
N/A
- TensorFlow installed from (source or binary):
binary
- TensorFlow version (use command below):
VERSION=1.13.0-dev20181225 (note: this is the 2.0-preview)
GIT_VERSION=b'v1.12.0-5131-gc6f3c5dc48'
- Python version:
3.6.6
- Bazel version (if compiling from source):
N/A
- GCC/Compiler version (if compiling from source):
N/A
- CUDA/cuDNN version:
N/A
- GPU model and memory:
N/A

**Describe the current behavior**
When evaluating a Keras model, the progress bar randomly stops before 100% (however, the loss and metrics returned by the function are correct). Also, it does not end with a newline.

**Describe the expected behavior**
I expect the progress bar to go up to 100% and display a newline.

**Code to reproduce the issue**

```python
import tensorflow as tf
import numpy as np

np.random.seed(42)
tf.random.set_seed(42)

(X_train, y_train), (X_test, y_test) = tf.keras.datasets.mnist.load_data()
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=[28, 28]),
    tf.keras.layers.Dense(10, activation="softmax"),
])
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd", metrics=["accuracy"])

model.fit(X_train, y_train, epochs=2)
print(model.evaluate(X_test, y_test))
```

**Other info / logs**
Here is the output of this program:

```
Epoch 1/2
60000/60000 [==============================] - 2s 28us/sample - loss: 32.8388 - acc: 0.8403
Epoch 2/2
60000/60000 [==============================] - 2s 27us/sample - loss: 26.3895 - acc: 0.8683
 9792/10000 [============================>.] - ETA: 0s - loss: 33.9531 - acc: 0.8363[33.969303797870886, 0.8358]
```

Notice that the evaluation progress bar (last line) does not go up to 100% (it stops at 9792/10000). Moreover, there is no newline at the end, so the function's returned values (`[33.969303797870886, 0.8358]`) are printed on the same line.

Moreover, when I run the same code again, I get a different output (only the last line differs). This time the progress bar stopped at 9088/10000, but notice that the function's results are the same as above:

```
Epoch 1/2
60000/60000 [==============================] - 2s 29us/sample - loss: 32.8388 - acc: 0.8403
Epoch 2/2
60000/60000 [==============================] - 2s 29us/sample - loss: 26.3895 - acc: 0.8683
 9088/10000 [==========================>...] - ETA: 0s - loss: 34.8416 - acc: 0.8327[33.969303797870886, 0.8358]
```

**Comment by yongtang** [comment]
I think the issue was that, `on_batch_end()` was relying on `on_epoch_end()` for the last step processing:
https://github.com/tensorflow/tensorflow/blob/8f60a381d210478f21762a6cf14f547a05e98878/tensorflow/python/keras/callbacks.py#L724-L727
but `on_epoch_end()` was only called for train mode:
https://github.com/tensorflow/tensorflow/blob/8f60a381d210478f21762a6cf14f547a05e98878/tensorflow/python/keras/engine/training_arrays.py#L375-L378

**Comment by yongtang** [comment]
Created a PR #24633 for the fix.

**Comment by ageron** [comment]
Once again, thanks @yongtang , and happy New Year! 👍

**Comment by yongtang** [comment]
Happy New Year🎉👍!

**Comment by lufol** [comment]
This issue can be closed, because some commit already fixed the problem according to  #24633.
